modules/contacts.js

```diff
@@ -0,0 +1,216 @@
+ */
+Contacts.prototype.backup = function() {
+  this.getContactsFromSIM();
+};
+
+/**
+ * @access public
+ * @description The Restore function will look for a contacts JSON file existing on an extenal SD card under the directory /backup/contacts/contacts.json. It will parse all the data in that file and transfer them back onto the internal memory in mozContacts
+ */
+Contacts.prototype.restore = function() {
+  var that = this;
+  var reader = new FileReader();
+
+  reader.onloadend = function() {
+    var contents = this.result;
+    var data = JSON.parse(contents);
+    for (var i = 0; i < data.length; ++i) {
+      var myContact = new mozContact(data[i]);
+      myContact.givenName = [data[i].name];
+      navigator.mozContacts.save(myContact);
+    }
+  };
+
+  var sdcard = navigator.getDeviceStorages('sdcard')[1];
+  var request = sdcard.get('/sdcard1/backup/contacts/contacts.json');
+
+  request.onsuccess = function() {
+    reader.readAsText(this.result);
+  };
+
+  request.onerror = function() {
+  };
+};
+
+/**
+ * @access public
+ * @description The clean function looks for a previous backup contacts.json file and will delete it if it exists as to not let the contacts functionality break when it tries to write its new backup to the external SD card
+ */
+Contacts.prototype.clean = function(oncomplete) {
+  var that = this;
+  var sdcard = navigator.getDeviceStorages('sdcard')[1];
+  var remove = sdcard.delete('/sdcard1/backup/contacts/contacts.json');
+
+  remove.onsuccess = function() {
+    if (window.ffosbr.utils.isFunction(oncomplete)) {
+      oncomplete('Clean success');
+    }
+  };
+
+  remove.onerror = function() {
+    if (window.ffosbr.utils.isFunction(oncomplete)) {
+      oncomplete(remove.error);
+    }
+  };
+};
+
+/**
+
+ * @access public
+ * @description This function gets contacts from the main memory where mozContacts are stored. It should only be called by the getContactsFromSIM function to avoid a race condition. Once all the contacts have been fetched by this function, it will call putContactsOnSD which stores the contacts from the SIM card and the internal memory to an external SD card.
+ */
+Contacts.prototype.getContactsFromOS = function() {
+  var that = this;
+  var allContactsCursor;
+
+  allContactsCursor = navigator.mozContacts.getAll({
+    sortBy: 'name',
+    sortOrder: 'ascending'
+  });
+
+  allContactsCursor.onsuccess = function() {
+    var contact = this.result;
+    if (contact) {
+      that.contacts.push(contact);
+      allContactsCursor.continue();
+    } else {
+      that.putContactsOnSD(function() {
+        //------Log what is written to the sdcard--------//
+        var sdcard = navigator.getDeviceStorages('sdcard')[1];
+        var cursor = sdcard.enumerate();
+        cursor.onsuccess = function() {
+          if (this.result) {
+            var file = this.result;
+            if (file.name === '/sdcard1/backup/contacts/contacts.json') {
```

Comment:
That is correct, I need dean to remove that code. Don't wanna touch it, but I think he said it was purely for debugging purposes.

Sent from my iPhone

> On Jul 22, 2015, at 10:13 PM, Ryan Bernstein <notifications@github.com> wrote:
> 
> In modules/contacts.js:
> 
> > +  });
> > +
> > +  allContactsCursor.onsuccess = function() {
> > +    var contact = this.result;
> > +    if (contact) {
> > +      that.contacts.push(contact);
> > +      allContactsCursor.continue();
> > +    } else {
> > +      that.putContactsOnSD(function() {
> > +        //------Log what is written to the sdcard--------//
> > +        var sdcard = navigator.getDeviceStorages('sdcard')[1];
> > +        var cursor = sdcard.enumerate();
> > +        cursor.onsuccess = function() {
> > +          if (this.result) {
> > +            var file = this.result;
> > +            if (file.name === '/sdcard1/backup/contacts/contacts.json') {
> Empty if body, unless I'm missing something?
> 
> —
> Reply to this email directly or view it on GitHub.
>